graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,372 @@
+    }
+
+    if (!executionResult.getErrors().isEmpty()) {
+      return super.instrumentExecutionResult(executionResult, parameters);
+    }
+
+    ExecutionResultImpl resultWithCacheControl =
+        ExecutionResultImpl.newExecutionResult()
+            .from(executionResult)
+            .addExtension(EXTENSION_KEY, state.getOverallPolicyString().get())
+            .build();
+
+    return CompletableFuture.completedFuture(resultWithCacheControl);
+  }
+
+  enum CacheControlScope {
+    PUBLIC,
+    PRIVATE
+  }
+
+  private static class CacheControlState implements InstrumentationState {
+    private final CacheControlPolicy overallPolicy = new CacheControlPolicy();
+
+    public Optional<String> getOverallPolicyString() {
+      return overallPolicy.getPolicy();
+    }
+
+    public void restrictCachePolicy(CacheControlPolicy policy) {
+      this.overallPolicy.restrict(policy);
+    }
+  }
+
+  private static class CacheControlPolicy implements InstrumentationState {
+    private Optional<Integer> maxAge = Optional.empty();
+    private Optional<CacheControlScope> scope = Optional.of(CacheControlScope.PUBLIC);
+
+    void restrict(CacheControlPolicy policy) {
+      if (policy.maxAge.isPresent()
+          && (!maxAge.isPresent() || policy.maxAge.get() < maxAge.get())) {
+        this.maxAge = policy.maxAge;
+      }
+
+      if (policy.scope.isPresent()
+          && (!scope.isPresent() || !scope.get().equals(CacheControlScope.PRIVATE))) {
+        this.scope = policy.scope;
+      }
+    }
+
+    void restrict(Hint hint) {
+      if (hint.maxAge.isPresent() && (!maxAge.isPresent() || hint.maxAge.get() < maxAge.get())) {
+        this.maxAge = hint.maxAge;
+      }
+
+      if (hint.scope.isPresent()
+          && (!scope.isPresent() || !scope.get().equals(CacheControlScope.PRIVATE))) {
+        this.scope = hint.scope;
+      }
+    }
+
+    void restrict(Integer maxAge) {
+      this.maxAge = Optional.of(maxAge);
+    }
+
+    void replace(Hint hint) {
+      if (hint.maxAge.isPresent()) {
+        this.maxAge = hint.maxAge;
+      }
+
+      if (hint.scope.isPresent()) {
+        this.scope = hint.scope;
+      }
+    }
+
+    void replace(Optional<CacheControlScope> scope) {
+      if (scope.isPresent()) {
+        this.scope = scope;
+      }
+    }
+
+    public Optional<String> getPolicy() {
+      Integer maxAgeValue = maxAge.orElse(0);
+      if (maxAgeValue.equals(0)) {
+        return Optional.empty();
+      }
+      return Optional.of(
+          String.format(
+              "max-age=%d, %s",
+              maxAgeValue, scope.orElse(CacheControlScope.PUBLIC).toString().toLowerCase()));
+    }
+
+    public boolean hasMaxAge() {
+      return maxAge.isPresent();
+    }
+
+    public boolean hasScope() {
+      return scope.isPresent();
+    }
+  }
+
+  private static class HintCache {
+    private final Map<GraphQLCompositeType, Hint> typeHintCache = new HashMap<>();
```

Comment:
I plead guilty of [suggesting using them as keys](https://github.com/apollographql/federation-jvm/pull/133#discussion_r748645320). I think that's ok?

Using a LruCache does sound like a good and safe idea 👍 . 

Maybe something as basic as 

```java
final int MAX_ENTRIES = 100;
Map cache = new LinkedHashMap(MAX_ENTRIES+1, .75F, true) {
    // This method is called just after a new entry has been added
    public boolean removeEldestEntry(Map.Entry eldest) {
        return size() > MAX_ENTRIES;
    }
};
```

I'm also assuming this might be multithreaded/parallelized so maybe worth locking it?

Reply:
thanks for the code sample! will refer back to this if we end up needing the cache. i definitely don't know how to estimate the size of the cache, so i'm just going to defer.